File: csv2graph.py
import argparse
import pandas as pd
from PIL import Image, ImageDraw
import colorsys
import math

# Usage example:
# python csv2graph.py --data ./hoge.csv --range 2 --columns a,c,e --out mygraph.png

def parse_args():
    parser = argparse.ArgumentParser(description="Generate scatter plot from CSV")
    parser.add_argument('--data', required=True, help="Path to CSV file")
    parser.add_argument('--range', type=float, help="Maximum value for X-axis (plot values less than or equal to this). If not specified, all data will be plotted")
    parser.add_argument('--columns', required=True, help="Comma-separated list of columns to plot")
    parser.add_argument('--size', default="768x512", help="Output image size in WIDTHxHEIGHT format (default is 768x512)")
    parser.add_argument('--skip', type=int, default=1, help="Data thinning interval (use only every Nth element, default is 1)")
    parser.add_argument('--xdata', type=lambda x: (str(x).lower() == 'true'), default=False,
                        help="Boolean indicating whether X-axis values are included in CSV (true or false, default is false)")
    parser.add_argument('--out', default="scatter_plot.png", help="Output image filename (default is scatter_plot.png)")
    parser.add_argument('--xscale', help="Map X-axis values to specified range (START,END format)")
    return parser.parse_args()

def generate_colors(n):
    colors = []
    for i in range(n):
        hue = i / n
        saturation = 0.7
        value = 0.9
        rgb = colorsys.hsv_to_rgb(hue, saturation, value)
        colors.append(tuple(int(x * 255) for x in rgb))
    return colors

def draw_grid(draw, width, height, margin, x_ticks, y_ticks):
    # Draw grid lines
    grid_color = (200, 200, 200)
    
    # X-axis grid lines (vertical lines)
    for x_pos in x_ticks:
        draw.line([(x_pos, margin), (x_pos, height - margin)], fill=grid_color, width=1)
    
    # Y-axis grid lines (horizontal lines)
    for y_pos in y_ticks:
        draw.line([(margin, y_pos), (width - margin, y_pos)], fill=grid_color, width=1)

def draw_axes(draw, width, height, margin, x_max, y_min, y_max, start):
    # Draw axes
    axis_color = (0, 0, 0)
    # X-axis
    draw.line([(margin, height - margin), (width - margin, height - margin)], fill=axis_color, width=2)
    # Y-axis
    draw.line([(margin, margin), (margin, height - margin)], fill=axis_color, width=2)
    
    # Calculate X-axis tick intervals
    available_width = width - 2 * margin
    digit_width = 20  # Estimated width per digit in pixels
    min_tick_spacing = 50  # Minimum spacing between ticks
    max_ticks = min(
        int(available_width / min_tick_spacing),
        int(available_width / (digit_width * 3))
    )
    x_step = max(1, int((x_max - start) / max_ticks))  # Calculate step considering start value
    
    # Calculate X-axis tick positions
    x_tick_positions = []
    x_tick_values = []
    
    for x in range(int(start), int(x_max) + 1, x_step):
        x_pos = margin + ((x - start) / (x_max - start)) * (width - 2 * margin)
        x_tick_positions.append(x_pos)
        x_tick_values.append(x)
    
    # Draw X-axis ticks
    for i, x_pos in enumerate(x_tick_positions):
        x = x_tick_values[i]
        draw.line([(x_pos, height - margin - 5), (x_pos, height - margin + 5)], fill=axis_color, width=1)
        draw.text((x_pos - 10, height - margin + 10), str(x), fill=axis_color)
    
    # Calculate Y-axis tick intervals and positions
    y_step = (y_max - y_min) / 5
    y_tick_positions = []
    
    for i in range(6):
        y = y_min + i * y_step
        y_pos = height - margin - (i / 5) * (height - 2 * margin)
        y_tick_positions.append(y_pos)
        draw.line([(margin - 5, y_pos), (margin + 5, y_pos)], fill=axis_color, width=1)
        draw.text((margin - 30, y_pos - 5), f"{y:.1f}", fill=axis_color)
    
    return x_tick_positions, y_tick_positions

def main():
    args = parse_args()

    # Load CSV file
    df = pd.read_csv(args.data)

    # Handle case where X-axis data doesn't exist
    if not args.xdata:
        df.insert(0, '_', range(1, len(df) + 1))

    # Get columns to plot
    columns_to_plot = args.columns.split(',')

    # Identify X-axis column (first column in CSV)
    x_col = df.columns[0]

    # Filter data by X-axis range
    if args.range is not None:
        df_filtered = df[df[x_col] <= args.range]
        x_max = args.range
    else:
        df_filtered = df
        x_max = df_filtered[x_col].max()

    # X-axis value range mapping
    start = 0
    end = x_max
    if args.xscale:
        start, end = map(float, args.xscale.split(','))
        # Set X-axis maximum to end
        x_max = end
        # Map values from start to end
        df_filtered[x_col] = start + (df_filtered[x_col] / df_filtered[x_col].max()) * (end - start)

    # Thin out data
    df_filtered = df_filtered.iloc[::args.skip, :]

    # Set image size
    width, height = map(int, args.size.split('x'))
    margin = 60  # Graph margin

    # Create new image
    image = Image.new('RGB', (width, height), 'white')
    draw = ImageDraw.Draw(image)

    # Calculate Y-axis range
    y_min = float('inf')
    y_max = float('-inf')
    for col in columns_to_plot:
        if col in df_filtered.columns:
            y_min = min(y_min, df_filtered[col].min())
            y_max = max(y_max, df_filtered[col].max())

    # Draw axes and get tick positions
    x_ticks, y_ticks = draw_axes(draw, width, height, margin, x_max, y_min, y_max, start)

    # Draw grid - aligned with X-axis ticks
    draw_grid(draw, width, height, margin, x_ticks, y_ticks)

    # Draw scatter plot
    colors = generate_colors(len(columns_to_plot))
    for i, col in enumerate(columns_to_plot):
        if col in df_filtered.columns:
            color = colors[i]
            # Sort data by X-axis value
            sorted_data = df_filtered.sort_values(by=x_col)
            points = []
            
            # Draw points and collect coordinates for lines
            for _, row in sorted_data.iterrows():
                x = row[x_col]
                y = row[col]
                # Map X-axis values
                x_pos = margin + (x - start) / (end - start) * (width - 2 * margin)
                y_pos = height - margin - ((y - y_min) / (y_max - y_min)) * (height - 2 * margin)
                points.append((x_pos, y_pos))
                draw.ellipse([(x_pos - 2, y_pos - 2), (x_pos + 2, y_pos + 2)], fill=color)
            
            # Connect points with lines
            if len(points) > 1:
                for j in range(len(points) - 1):
                    draw.line([points[j], points[j + 1]], fill=color, width=1)

    # Position legend outside the graph area (top-right)
    legend_padding = 10  # Internal padding for legend
    legend_box_size = 10
    legend_text_offset = 15
    legend_spacing = 20
    
    # Calculate legend width dynamically (based on column name length)
    valid_columns = [col for col in columns_to_plot if col in df_filtered.columns]
    max_column_length = max([len(col) for col in valid_columns], default=0)
    char_width = 7  # Estimated width per character (pixels)
    text_width = max_column_length * char_width
    legend_width = legend_box_size + legend_text_offset + text_width + 2 * legend_padding
    
    # Calculate legend height
    legend_height = len(valid_columns) * legend_spacing + 2 * legend_padding
    
    # Position legend outside graph area (top-right)
    legend_left = width - margin + 10  # Right edge of graph area + 10px margin
    legend_top = margin  # Same height as top of graph area
    
    # Adjust to prevent exceeding right edge of image
    if legend_left + legend_width > width:
        legend_left = width - legend_width - 5  # 5px margin from edge
    
    # Draw legend background
    draw.rectangle(
        [(legend_left, legend_top), (legend_left + legend_width, legend_top + legend_height)],
        fill=(255, 255, 255),  # Background color
        outline=(100, 100, 100)
    )
    
    # Draw legend items
    legend_y_current = legend_top + legend_padding
    for i, col in enumerate(valid_columns):
        color = colors[i]
        draw.rectangle(
            [(legend_left + legend_padding, legend_y_current), 
             (legend_left + legend_padding + legend_box_size, legend_y_current + legend_box_size)],
            fill=color
        )
        draw.text(
            (legend_left + legend_padding + legend_text_offset, legend_y_current), 
            col, 
            fill=(0, 0, 0)
        )
        legend_y_current += legend_spacing

    # Draw title
    draw.text((width // 2 - 100, 20), 'Scatter Plot from CSV', fill=(0, 0, 0))

    # Generate output image file
    output_image = args.out
    image.save(output_image)
    print(f"Graph saved to {output_image}")

if __name__ == "__main__":
    main()

File: csv2graph.go
package main

import (
	"encoding/csv"
	"flag"
	"fmt"
	"image/color"
	"math"
	"os"
	"strconv"
	"strings"

	"github.com/fogleman/gg"
)

func main() {
	// Parse command line arguments
	dataPath := flag.String("data", "", "Path to CSV file")
	maxRange := flag.Float64("range", -1, "Max X value, plot points <= this (optional)")
	columns := flag.String("columns", "", "Comma-separated columns to plot")
	size := flag.String("size", "768x512", "Output image size (e.g. 768x512)")
	skip := flag.Int("skip", 1, "Data thinning (plot every Nth point, default=1)")
	xdata := flag.Bool("xdata", false, "When specified, CSV has X-axis values in first column")
	out := flag.String("out", "scatter_plot.png", "Output image filename")
	xscale := flag.String("xscale", "", "Map X values to range START,END")
	title := flag.String("title", "Scatter Plot from CSV", "Graph title")
	flag.Parse()

	if *dataPath == "" || *columns == "" {
		fmt.Println("Usage: csv2graph --data <FILE> --columns <COLS> [options]")
		return
	}

	colsToPlot := strings.Split(*columns, ",")

	// Open CSV file
	f, err := os.Open(*dataPath)
	if err != nil {
		fmt.Println("CSV open error:", err)
		return
	}
	defer f.Close()

	csvReader := csv.NewReader(f)
	records, err := csvReader.ReadAll()
	if err != nil {
		fmt.Println("CSV read error:", err)
		return
	}
	if len(records) < 2 {
		fmt.Println("No data to plot")
		return
	}

	// Parse header and data rows
	header := records[0]
	dataRows := records[1:]

	// Generate X column if xdata=false
	var xIndex int
	if !*xdata {
		// Insert X column
		header = append([]string{"_"}, header...)
		for i := 0; i < len(dataRows); i++ {
			dataRows[i] = append([]string{strconv.Itoa(i + 1)}, dataRows[i]...)
		}
	}
	xIndex = 0

	// Map column names to indexes
	colIndexMap := make(map[string]int)
	for i, h := range header {
		colIndexMap[h] = i
	}

	// Filter data by range
	var filtered [][]string
	if *maxRange > 0 {
		for _, row := range dataRows {
			xVal, err := strconv.ParseFloat(row[xIndex], 64)
			if err == nil && xVal <= *maxRange {
				filtered = append(filtered, row)
			}
		}
	} else {
		filtered = dataRows
	}

	// Get original data range
	origXMin, origXMax := math.Inf(1), math.Inf(-1)
	if len(filtered) > 0 {
		for _, row := range filtered {
			xVal, err := strconv.ParseFloat(row[xIndex], 64)
			if err == nil {
				if xVal < origXMin {
					origXMin = xVal
				}
				if xVal > origXMax {
					origXMax = xVal
				}
			}
		}
	}

	// Determine plot display range
	plotXMin, plotXMax := origXMin, origXMax
	if *xscale != "" {
		parts := strings.Split(*xscale, ",")
		if len(parts) == 2 {
			s, errS := strconv.ParseFloat(parts[0], 64)
			e, errE := strconv.ParseFloat(parts[1], 64)
			if errS == nil && errE == nil && e > s {
				plotXMin, plotXMax = s, e
			}
		}
	}

	// Data thinning
	thinned := [][]string{}
	for i := 0; i < len(filtered); i += *skip {
		thinned = append(thinned, filtered[i])
	}

	// Parse image size
	wh := strings.Split(*size, "x")
	w, _ := strconv.Atoi(wh[0])
	h, _ := strconv.Atoi(wh[1])

	// Prepare drawing context
	dc := gg.NewContext(w, h)
	dc.SetColor(color.White)
	dc.Clear()

	// Find Y-axis min, max values
	yMin, yMax := math.Inf(1), math.Inf(-1)
	for _, c := range colsToPlot {
		idx, ok := colIndexMap[c]
		if !ok {
			continue
		}
		for _, row := range thinned {
			val, err := strconv.ParseFloat(row[idx], 64)
			if err == nil {
				if val < yMin {
					yMin = val
				}
				if val > yMax {
					yMax = val
				}
			}
		}
	}
	if yMin == math.Inf(1) || yMax == math.Inf(-1) {
		fmt.Println("No valid data for plotting")
		return
	}

	// Margin and usable area
	margin := 60.0
	usableW := float64(w) - 2*margin
	usableH := float64(h) - 2*margin

	// Draw axes
	dc.SetColor(color.Black)
	dc.SetLineWidth(2)
	// X-axis
	dc.DrawLine(margin, float64(h)-margin, float64(w)-margin, float64(h)-margin)
	dc.Stroke()
	// Y-axis
	dc.DrawLine(margin, margin, margin, float64(h)-margin)
	dc.Stroke()

	// X-axis numeric labels
	xTickSteps := 5
	for i := 0; i <= xTickSteps; i++ {
		// Calculate label value
		labelVal := plotXMin + (float64(i)/float64(xTickSteps))*(plotXMax-plotXMin)
		// Calculate label position
		tx := margin + (float64(i)/float64(xTickSteps))*usableW
		dc.SetColor(color.Black)
		dc.DrawStringAnchored(fmt.Sprintf("%.1f", labelVal), tx, float64(h)-margin+15, 0.5, 0)
	}

	// Y-axis numeric labels
	yTickSteps := 5
	for i := 0; i <= yTickSteps; i++ {
		labelVal := yMin + (float64(i)/float64(yTickSteps))*(yMax-yMin)
		ty := float64(h) - margin - (float64(i)/float64(yTickSteps))*usableH
		dc.SetColor(color.Black)
		dc.DrawStringAnchored(fmt.Sprintf("%.1f", labelVal), margin-10, ty, 1, 0.5)
	}

	// Draw grid
	for i := 0; i <= xTickSteps; i++ {
		// X-axis grid
		tx := margin + (float64(i)/float64(xTickSteps))*usableW
		dc.SetColor(color.RGBA{200, 200, 200, 255})
		dc.DrawLine(tx, margin, tx, float64(h)-margin)
		dc.Stroke()
	}

	for i := 0; i <= yTickSteps; i++ {
		// Y-axis grid
		ty := float64(h) - margin - (float64(i)/float64(yTickSteps))*usableH
		dc.SetColor(color.RGBA{200, 200, 200, 255})
		dc.DrawLine(margin, ty, float64(w)-margin, ty)
		dc.Stroke()
	}

	// Plot each column
	colors := generateColors(len(colsToPlot))
	for cidx, c := range colsToPlot {
		idx, ok := colIndexMap[c]
		if !ok {
			continue
		}
		points := make([][2]float64, 0)
		for _, row := range thinned {
			xval, err1 := strconv.ParseFloat(row[xIndex], 64)
			yval, err2 := strconv.ParseFloat(row[idx], 64)
			if err1 == nil && err2 == nil {
				// Normalize X value if xscale is specified
				normalizedX := xval
				if *xscale != "" {
					// Scale the original X value to match the specified range
					// Original: [origXMin, origXMax] -> Target: [plotXMin, plotXMax]
					normalizedX = plotXMin + ((xval-origXMin)/(origXMax-origXMin))*(plotXMax-plotXMin)
				}

				// Calculate screen coordinates using normalized X
				xx := margin + ((normalizedX-plotXMin)/(plotXMax-plotXMin))*usableW

				// Y value mapping: data range -> screen coordinates
				yy := float64(h) - margin - ((yval-yMin)/(yMax-yMin))*usableH

				// Plot only data within display area
				if xx >= margin && xx <= float64(w)-margin {
					points = append(points, [2]float64{xx, yy})
				}
			}
		}

		dc.SetColor(colors[cidx])
		dc.SetLineWidth(1) // Set thin line width
		for i, p := range points {
			// Draw point
			dc.DrawCircle(p[0], p[1], 2)
			dc.Fill()
			// Draw line (from second point)
			if i > 0 {
				dc.DrawLine(points[i-1][0], points[i-1][1], p[0], p[1])
				dc.Stroke()
			}
		}
	}

	// Draw legend
	legendX := float64(w) - margin + 10
	legendY := margin
	boxSize := 10.0
	spacing := 20.0
	dc.SetColor(color.White)
	lw := 100.0
	lh := float64(len(colsToPlot))*spacing + 20
	dc.DrawRectangle(legendX, legendY, lw, lh)
	dc.Fill()
	dc.SetColor(color.Black)
	dc.DrawRectangle(legendX, legendY, lw, lh)
	dc.Stroke()

	yy := legendY + 10
	for i, c := range colsToPlot {
		dc.SetColor(colors[i])
		dc.DrawRectangle(legendX+10, yy, boxSize, boxSize)
		dc.Fill()
		dc.SetColor(color.Black)
		dc.DrawStringAnchored(c, legendX+30, yy+boxSize/2, 0, 0.5)
		yy += spacing
	}

	// Title
	dc.SetColor(color.Black)
	dc.DrawStringAnchored(*title, float64(w)/2, 20, 0.5, 0.5)

	// Save
	outF, err := os.Create(*out)
	if err != nil {
		fmt.Println("Create file error:", err)
		return
	}
	defer outF.Close()
	dc.EncodePNG(outF)
	fmt.Println("Saved:", *out)
}

func generateColors(n int) []color.Color {
	var cols []color.Color
	for i := 0; i < n; i++ {
		// HSV -> RGB
		h := float64(i) / float64(n)
		s := 0.7
		v := 0.9
		r, g, b := hsv2rgb(h, s, v)
		cols = append(cols, color.RGBA{uint8(r * 255), uint8(g * 255), uint8(b * 255), 255})
	}
	return cols
}

func hsv2rgb(h, s, v float64) (float64, float64, float64) {
	var r, g, b float64
	i := math.Floor(h * 6)
	f := h*6 - i
	p := v * (1 - s)
	q := v * (1 - f*s)
	t := v * (1 - (1-f)*s)
	switch int(i) % 6 {
	case 0:
		r, g, b = v, t, p
	case 1:
		r, g, b = q, v, p
	case 2:
		r, g, b = p, v, t
	case 3:
		r, g, b = p, q, v
	case 4:
		r, g, b = t, p, v
	case 5:
		r, g, b = v, p, q
	}
	return r, g, b
}
